lib/thesis/controller.ex

```diff
@@ -23,6 +23,53 @@ defmodule Thesis.Controller do
   defmacro __using__(_) do
     quote do
       plug Thesis.Controller.Plug
+
+      @doc """
+      Renders a page that exists only in the database, as opposed to a page that
+      is defined in the router.
+
+      This function will also redirect to the page's `redirect_url` if it is set.
+      However, if the user is in edit mode, we will render the default template and
+      display an alert telling them about the redirect. This gives the user a chance
+      to edit where this page redirects to.
+      """
+      def render_dynamic(conn, opts \\ []) do
+        page = conn.assigns[:thesis_page]
+        if page && page.id do
+          do_render_dynamic(conn, page, opts)
+        else
+          render_not_found(conn)
+        end
+      end
+
+      defp do_render_dynamic(conn, page, opts) do
+        if Thesis.Page.redirected?(page) && !conn.assigns[:thesis_editable] do
+          conn
+          |> put_status(301)
+          |> redirect(to: page.redirect_url)
+        else
+          conn
+          |> put_status(200)
+          |> put_view(Thesis.Config.dynamic_view || conn.private.phoenix_view)
+          |> render(page_template(page, opts))
+        end
+      end
```

Comment:
If the user chose to put a redirect_url, then let's 301 redirect to the new url. However, if we're in edit mode (e.g. logged in and able to edit this Thesis page), then we don't automatically redirect, since we need the opportunity to edit where this page redirects or even remove the redirect entirely. I'll show how that's done later.